eservice-versions-archiver/src/main/scala/it/pagopa/interop/eserviceversionsarchiver/util/JobExecution.scala

```diff
@@ -0,0 +1,66 @@
+package it.pagopa.interop.eserviceversionsarchiver.util
+
+import io.circe.parser._
+import it.pagopa.interop.agreementmanagement.model.agreement.{Archived, PersistentAgreement}
+import it.pagopa.interop.agreementmanagement.model.persistence.JsonFormats.paFormat
+import it.pagopa.interop.catalogmanagement.model.persistence.JsonFormats.ciFormat
+import it.pagopa.interop.catalogmanagement.model.{CatalogItem, Deprecated, Published, Suspended}
+import it.pagopa.interop.commons.cqrs.service.MongoDbReadModelService
+import it.pagopa.interop.commons.utils.TypeConversions._
+import it.pagopa.interop.eserviceversionsarchiver.ApplicationConfiguration._
+import it.pagopa.interop.eserviceversionsarchiver.service.CatalogProcessService
+import it.pagopa.interop.eserviceversionsarchiver.util.ReadModelQueries.getAllAgreements
+import it.pagopa.interop.eserviceversionsarchiver.util.errors._
+import org.mongodb.scala.model.Filters
+import spray.json.BasicFormats
+
+import scala.concurrent.Future
+
+final case class JobExecution(readModelService: MongoDbReadModelService, catalogProcess: CatalogProcessService)
+    extends BasicFormats {
+
+  def archiveEservice(message: String): Future[Unit] =
+    for {
+      json        <- parse(message).toFuture
+      agreementId <- json.hcursor.get[String]("agreementId").toFuture
+      _           <- archiveEserviceInternal(agreementId)
+    } yield ()
+
+  private def archiveEserviceInternal(agreementId: String): Future[Unit] = {
+    for {
+      maybeAgreement             <- readModelService
+        .findOne[PersistentAgreement](agreementsCollection, Filters.eq("data.id", agreementId))
+      (eServiceId, descriptorId) <- maybeAgreement
+        .toFuture(AgreementNotFound(agreementId))
+        .map(agreement => (agreement.eserviceId, agreement.descriptorId))
+      descriptorAndEserviceFilter = Filters.and(
+        Filters.eq("data.descriptorId", descriptorId),
+        Filters.eq("data.eserviceId", eServiceId)
+      )
+      relatingAgreements <- getAllAgreements(readModelService, descriptorAndEserviceFilter)
+      _                  <- {
+        val allArchived = relatingAgreements.forall(_.state == Archived)
+        if (allArchived) Future.unit
+        else Future.failed(NonArchivableDescriptorException(eServiceId, descriptorId))
+      }
```

Comment:
any reason why this should fail?

Take this use case:
- Agreements A1 and A2 exist, both are ACTIVE and both are on EService E1 and Descriptor D1
- Agreement A1 is archived, but A2 is not

this is an acceptable state.

A job should fail with error only if it encounters an error/unexpected condition.
In this case this message is simply out of scope of this job, and can be ignored, it does not produce a descriptor archiviation and the job should successfully move to the next message

Reply:
You're right, in this case how can I break out of the for comprehension when the execution meets a condition that makes a Descriptor non-archiveable? Because if just one of these condition is met it's useless to continue checking for the others.

I thought that returning Future.failed it would exit from the for-comprehension and execute it with the next message

Reply:
Using the failure to short-circuit the logic is possible but not advisable in this case, because the failure should represent an `exception` to the logic, while this is part of the logic itself.
It would be like throwing an exception just to exit from a loop.

One approach is to enclose the logic in a function and call it only when the conditions are met:
```
      relatingAgreements <- getAllAgreements(readModelService, descriptorAndEserviceFilter)
      allArchived = relatingAgreements.forall(_.state == Archived)
      _                  <-  if (allArchived) archiveEServiceDescriptor(...)
                                else Future.unit
```
(`cats` has nice sugar functions for these cases, but unfortunately we cannot use them on `Future`)

Reply:
The `archiveEServiceDescriptor` function would be just `catalogProcess.archiveDescriptor(eServiceId, descriptorId)`

Wouldn't, in this case, continue the execution of the rest of the for-comprehension when we return `Future.unit`?

In order to stop the for-comprehension shouldn't we use something like the following (I'm not saying that this is good 👀) ?

      relatingAgreements <- getAllAgreements(readModelService, descriptorAndEserviceFilter)
      allArchived = relatingAgreements.forall(_.state == Archived)
      _                  <-  if (allArchived) catalogProcess.archiveDescriptor(eServiceId, descriptorId)
                                else return

Reply:
`archiveEServiceDescriptor` would not contain only `catalogProcess.archiveDescriptor` but all the logic required to archive the descriptor